operators/pkg/tenant-controller/tenant_controller.go

```diff
@@ -466,3 +396,185 @@ func updateTnNetPolAllow(np *netv1.NetworkPolicy) {
+		return err
+	}
+	if userID == nil {
+		userID, err = r.KcA.createKcUser(ctx, tn.Name, tn.Spec.FirstName, tn.Spec.LastName, tn.Spec.Email)
+	} else {
+		err = r.KcA.updateKcUser(ctx, *userID, tn.Spec.FirstName, tn.Spec.LastName, tn.Spec.Email, *currentUserEmail != tn.Spec.Email)
+	}
+	if err != nil {
+		klog.Errorf("Error when creating or updating keycloak user %s -> %s", tn.Name, err)
+		return err
+	} else if err = r.KcA.updateUserRoles(ctx, genKcUserRoleNames(tenantExistingWorkspaces), *userID, "workspace-"); err != nil {
+		klog.Errorf("Error when updating user roles of user %s -> %s", tn.Name, err)
+		return err
+	}
+	klog.Infof("Keycloak resources of user %s updated", tn.Name)
+	return nil
+}
+
+// genKcUserRoleNames maps the workspaces of a tenant to the needed roles in keycloak
+func genKcUserRoleNames(workspaces []crownlabsv1alpha1.UserWorkspaceData) []string {
+	userRoles := make([]string, len(workspaces))
+	// convert workspaces to actual keyloak role
+	for i, ws := range workspaces {
+		userRoles[i] = fmt.Sprintf("workspace-%s:%s", ws.WorkspaceRef.Name, ws.Role)
+	}
+	return userRoles
+}
+
+func (r TenantReconciler) handleNextcloudSubscription(ctx context.Context, tn *crownlabsv1alpha1.Tenant, nsName string) error {
+	// independently of the existence of the nexctloud secret for the nextcloud credentials of the user, need to know the displayname of the user, in order to update it if necessary
+	ncUsername := genNcUsername(tn.Name)
+	expectedDisplayname := genNcDisplayname(tn.Spec.FirstName, tn.Spec.LastName)
+	ncSecret := v1.Secret{ObjectMeta: metav1.ObjectMeta{Name: "nextcloud-credentials", Namespace: nsName}}
+	ncUserFound, ncDisplayname, err := r.NcA.GetUser(ncUsername)
+	switch {
+	case err != nil:
+		klog.Errorf("Error when getting nextcloud user: %s", err)
+		return err
+	case ncUserFound:
+		if err = r.Get(ctx, types.NamespacedName{Name: ncSecret.Name, Namespace: ncSecret.Namespace}, &ncSecret); client.IgnoreNotFound(err) != nil {
+			klog.Errorf("Error when getting nextcloud secret for tenant %s: %s", tn.Name, err)
+			return err
+		} else if err != nil {
+			ncPsw, err := generateToken()
+			if err != nil {
+				klog.Errorf("Error when generating nextcloud password of tenant %s: %s", tn.Name, err)
+				return err
+			}
+			if err := r.NcA.UpdateUserData(ncUsername, "password", *ncPsw); err != nil {
+				klog.Errorf("Error when updating password of tenant %s: %s", tn.Name, err)
+				return err
+			}
+			// nextcloud secret not found, need to create it
+			updateTnNcSecret(&ncSecret, ncUsername, *ncPsw)
+			if err := ctrl.SetControllerReference(tn, &ncSecret, r.Scheme); err != nil {
+				klog.Errorf("Error when adding controller reference to nextcloud secret for tenant %s: %s", tn.Name, err)
+				return err
+			} else if err = r.Create(ctx, &ncSecret); err != nil {
+				klog.Errorf("Error when creating nextcloud secret for tenant %s: %s", tn.Name, err)
+				return err
+			} else {
+				return nil
+			}
```

Comment:
Just curiosity. Why aren't you using the CreateOrUpdate function as always?

Reply:
Because, in theory, some lines above the operator tries to get the secret and continues only if the secret doesn't exist (it means that the password must be changed) so only a create is needed

Reply:
Ok, I see. I would honestly keep the other approach for two reasons: consistency with the rest of the code and avoid strange failures if the secret is already present (although it should never happen)